pubsys: update to AWS SDK Rust

jpculp opened the pull request:
**Issue number:**

#1968

**Description of changes:**

Replaces **rusoto** with **aws-sdk-for-rust** in **pubsys**.

**Testing done:**
- [x] publish ami to us-west-2
- [x] publish multiple amis with chained assumed roles (2 different accounts, 3 regions)
- [x] grant ami access
- [x] revoke ami access
- [x] publish ssm parameters
- [x] promote ssm parameters

**TUF testing done:**
_Thanks, @etungsten!_
- [x] use KMS key when creating TUF repo
- [x] check TUF repo expirations
- [x] validate TUF repo
- [x] refresh TUF repo

**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

jpculp commented on tools/pubsys/src/aws/client.rs:
```diff
@@ -1,161 +1,164 @@
+    if maybe_regional_role.is_some() {
+        build_chained_config(&region, sts_region, assume_roles.clone(), &aws.profile)
+            .await
+            .expect("?????")
```
This needs a proper expect message or an alternative method of unwrapping the result.

jpculp replied:
Fixed this in my latest push.

etungsten commented on tools/pubsys/src/aws/ami/mod.rs:
```diff
@@ -119,77 +118,58 @@ async fn _run(args: &Args, ami_args: &AmiArgs) -> Result<HashMap<String, Image>>
     // We register in this base region first, then copy from there to any other regions.
     let base_region = regions.remove(0);
 
+    let arch = ami_args.arch.as_str().to_owned();
```
nit: I would drop the `.to_owned()` here and do that for the places that need it.
```suggestion
    let arch = ami_args.arch.as_str();
```

Then when you're passing the `arch` to the error enum variants, you can just for example do:
```
...
    .context(error::GetAmiIdSnafu {
        name: &ami_args.name,
        arch,
        region: &base_region,
    })?;
```

Same comment applies elsewhere where you do similar things with `let arch`.

jpculp replied:
I was running into borrow issues earlier, but it looks like they aren't there now.

jpculp commented:
Fixed the result unwrapping issue while also improving how and where the profile gets loaded.

zmrow commented on tools/pubsys/src/aws/client.rs:
```diff
@@ -1,159 +1,151 @@
-}
+    // Add custom endpoint if specified in aws.region.REGION.endpoint.
+    let maybe_regional_endpoint = aws.region.get(&region).and_then(|r| r.endpoint.clone());
+    config = if let Some(endpoint) = &maybe_regional_endpoint {
```
You probably don't need the additional `config =` assignment here or the `else` (unless I'm missing something).

Since `config` is mutable, you could do:
```
if let Some(...) = ... {
    config.endpoint_resolver...
}
```

zmrow replied:
If `config.endpoint_resolver` isn't a "setter" I suppose my suggestion won't work.

zmrow replied:
Ah - according to the docs it consumes `self` and overrides the endpoint resolver. https://docs.rs/aws-config/latest/aws_config/struct.ConfigLoader.html#method.endpoint_resolver

zmrow commented on tools/pubsys/src/aws/client.rs:
```diff
@@ -1,159 +1,151 @@
-}
+    // If the user specified a profile, use that, otherwise use the default
+    // credentials mechanisms.
+    config = if let Some(profile) = &aws.profile {
```
Same here.

zmrow commented on tools/pubsys/src/aws/client.rs:
```diff
@@ -1,159 +1,151 @@
-    }
+    // Load a chained credential config if role is specified in aws.region.REGION.role.
+    let maybe_regional_role = aws.region.get(&region).and_then(|r| r.role.clone());
+    let config = if maybe_regional_role.is_some() {
```
I'm confused about this `let config`... are we replacing the `config` we just built?

jpculp replied:
Correct, but I can rename it to make it more clear.

stmcginnis reviewed (commented):
This looks good to me - with the caveat that I am not very familiar with this code. But updated code looks logical, and maybe most importantly, all tests are passing.

zmrow reviewed (commented):
As an aside, if you're not already doing so it'd probably be good to run down the entire set of pubsys commands in your own account, making sure that the grant/revoke and SSM params all work as expected.  Those are super important.

jpculp commented:
- Fixed missing region from chained configuration.
- Removed the extraneous `else`'s.

etungsten commented on tools/pubsys/src/repo.rs:
```diff
@@ -753,6 +763,13 @@ mod error {
+        #[snafu(display("Unable to create tokio runtime: {}", source))]
+        RuntimeCreation {
+            source: std::io::Error,
+            backtrace: Backtrace,
```
Why do we need a backtrace here? I also think you need to enable the `backtraces` feature flag for the snafu crate for this to have any effect? https://docs.rs/snafu/0.6.3/snafu/guide/feature_flags/index.html#backtraces

jpculp replied:
I only had it there to stay consistent with our implementation in [`tough-kms`](https://github.com/awslabs/tough/blob/develop/tough-kms/src/error.rs#L19), but removing it doesn't cause any issues that I'm aware of.

etungsten commented on tools/pubsys/src/main.rs:
```diff
@@ -39,8 +40,17 @@ fn run() -> Result<()> {
+        // Set log level for AWS SDK to error to reduce verbosity.
+        LevelFilter::Info => Builder::new()
+            .filter_level(args.log_level)
+            .filter(Some("aws_config"), LevelFilter::Error)
+            .filter(Some("aws_smithy"), LevelFilter::Error)
```
What does filtering at `Warning` look like for the `aws-sdk-rust` crates? Is it just as noisy?

jpculp replied:
I set to filter at `Error` since that's the level that was set in `bottlerocket-test-system`, but `Warn` could probably make sense here.

etungsten commented on tools/pubsys/src/main.rs:
```diff
@@ -39,8 +40,17 @@ fn run() -> Result<()> {
+            .filter(Some("aws_smithy"), LevelFilter::Error)
+            .init(),
+
+        // Set the supplied log level across the whole crate.
+        _ => Builder::new().filter_level(args.log_level).init(),
```
If I don't explicitly set the logging level for `pubsys`, I think I'd still want to skip out on all the extraneous `info` level log from the `aws-rust-sdk` crates.

jpculp replied:
Since `pubsys` will get a default value of `INFO` for `args.log_level`, I think this works as is. The filtering will apply even if the level isn't explicitly set, and if one were passed we would want that intentional log level applied to everything.

etungsten commented on tools/pubsys/src/aws/ssm/ssm.rs:
```diff
@@ -97,7 +94,7 @@ where
             total_count == expected_len,
             error::MissingInResponseSnafu {
-                region: region.name(),
+                region: &region,
```
nit: can we do this?:

```suggestion
                region,
```

etungsten commented on tools/pubsys/src/aws/publish_ami/mod.rs:
```diff
@@ -503,31 +494,30 @@ pub(crate) async fn modify_regional_images(
-        Client {
-            client_type: String,
+        #[snafu(display("Error creating client config in {}: {}", region, source))]
+        ClientConfig {
```
We removed the `client_type` field from this error. When something does fail with the client config generation, is it clear which type of client (e.g. EC2 client) we're generating the config for? Or does it not matter?

jpculp replied:
It doesn't matter, but I'll throw it back in to be closer to the original messaging.

etungsten commented on tools/pubsys/src/aws/publish_ami/mod.rs:
```diff
@@ -150,21 +147,21 @@ pub(crate) async fn run(args: &Args, publish_args: &PublishArgs) -> Result<()> {
         wait_response.context(error::WaitAmiSnafu {
             id: &image_id,
-            region: region.name(),
+            region: &region,
```
nit: can we do this?
```suggestion
            region,
```

etungsten commented on tools/pubsys/src/aws/publish_ami/mod.rs:
```diff
@@ -120,28 +119,26 @@ pub(crate) async fn run(args: &Args, publish_args: &PublishArgs) -> Result<()> {
     );
 
-    // Parse region names, adding endpoints from InfraConfig if specified
+    // Parse regions, adding endpoints from InfraConfig if specified
```
Are we still parsing the region string? 
```suggestion
    // Adding endpoints from InfraConfig if specified
```

etungsten commented on tools/pubsys/src/aws/client.rs:
```diff
@@ -1,159 +1,150 @@
+                    role_arn: &assume_role,
+                })?,
+            sts_credentials.session_token().map(|s| s.to_string()),
+            Some(
+                SystemTime::try_from(sts_credentials.expiration.unwrap())
+                    .context(error::DateTimeToSystemTimeSnafu)?,
+            ),
```
Is that a safe `unwrap`? What if it's `None`? Why can't we use the `sts_credentials.expiration` as is?

etungsten commented on tools/pubsys/src/aws/client.rs:
```diff
@@ -1,159 +1,150 @@
+    let maybe_regional_endpoint = aws.region.get(&region).and_then(|r| r.endpoint.clone());
+    if let Some(endpoint) = &maybe_regional_endpoint {
+        config =
+            config.endpoint_resolver(Endpoint::immutable(endpoint.parse().expect("valid URI")));
```
Let's avoid panicking if at all possible.
```suggestion
            config.endpoint_resolver(Endpoint::immutable(endpoint.parse().context(...)?));
```

etungsten reviewed (changes requested):
Can you please also test `cargo make check-repo-expirations`, `cargo make validate-repo`, and `cargo make refresh-repo`?

You can find examples in the PR descriptions here:
https://github.com/bottlerocket-os/bottlerocket/pull/2303
https://github.com/bottlerocket-os/bottlerocket/pull/1575

jpculp commented:
- Removed extraneous clones and references.
- Set default log level for `aws_config` and `aws_smithy` to `Warn`.
- Improved error handling and logging.

etungsten commented:
Tested check-repo-expirations:
```
$ cargo make  -e BUILDSYS_VARIANT=aws-k8s-1.23 -e BUILDSYS_ARCH=x86_64 check-repo-expirations
[cargo-make] INFO - cargo make 0.35.15
[cargo-make] INFO - Build File: Makefile.toml
[cargo-make] INFO - Task: check-repo-expirations
[cargo-make] INFO - Profile: development
[cargo-make] INFO - Running Task: setup
[cargo-make] INFO - Running Task: fetch-sources
[cargo-make] INFO - Running Task: tuftool
[cargo-make] INFO - Running Task: publish-setup-tools
[cargo-make] INFO - Running Task: publish-setup
19:01:24 [INFO] Found infra config at path: /home/ec2-user/bottlerocket/Infra.toml
[cargo-make] INFO - Running Task: setup
[cargo-make] INFO - Running Task: fetch-sources
[cargo-make] INFO - Running Task: publish-tools
[cargo-make] INFO - Running Task: check-repo-expirations
[2022-09-19T19:01:26Z INFO  pubsys_config] Found infra config at path: /home/ec2-user/bottlerocket/Infra.toml
[2022-09-19T19:01:26Z INFO  pubsys::repo::check_expirations] Loaded TUF repo:   file:///home/ec2-user/bottlerocket/build/repos/default/bottlerocket-1.9.2-deea371c/aws-k8s-1.23/x86_64
[2022-09-19T19:01:26Z INFO  pubsys::repo::check_expirations] Root expiration:   2023-09-18 18:58:32 UTC
[2022-09-19T19:01:26Z INFO  pubsys::repo::check_expirations] Snapshot expiration:       2022-10-03 19:01:12.156262570 UTC
[2022-09-19T19:01:26Z INFO  pubsys::repo::check_expirations] Targets expiration:        2022-10-03 19:01:12.156262570 UTC
[2022-09-19T19:01:26Z INFO  pubsys::repo::check_expirations] Timestamp expiration:      2022-09-26 19:01:12.156262570 UTC
[2022-09-19T19:01:26Z INFO  pubsys::repo::check_expirations] Looking for metadata expirations happening from now to 2022-09-22 19:01:26.973914183 UTC
[cargo-make] INFO - Build Done in 4.72 seconds.
```

Tested `validate-repo`:
```
$ cargo make  -e BUILDSYS_VARIANT=aws-k8s-1.23 -e BUILDSYS_ARCH=x86_64 validate-repo
[cargo-make] INFO - cargo make 0.35.15
[cargo-make] INFO - Build File: Makefile.toml
[cargo-make] INFO - Task: validate-repo
[cargo-make] INFO - Profile: development
[cargo-make] INFO - Running Task: setup
[cargo-make] INFO - Running Task: fetch-sources
[cargo-make] INFO - Running Task: tuftool
[cargo-make] INFO - Running Task: publish-setup-tools
[cargo-make] INFO - Running Task: publish-setup
19:01:45 [INFO] Found infra config at path: /home/ec2-user/bottlerocket/Infra.toml
[cargo-make] INFO - Running Task: setup
[cargo-make] INFO - Running Task: fetch-sources
[cargo-make] INFO - Running Task: publish-tools
[cargo-make] INFO - Running Task: validate-repo
[2022-09-19T19:01:47Z INFO  pubsys_config] Found infra config at path: /home/ec2-user/bottlerocket/Infra.toml
[2022-09-19T19:01:47Z INFO  pubsys::repo::validate_repo] Loaded TUF repo: file:///home/ec2-user/bottlerocket/build/repos/default/bottlerocket-1.9.2-deea371c/aws-k8s-1.23/x86_64
[2022-09-19T19:01:47Z INFO  pubsys::repo::validate_repo] Downloading target: migrate_v1.10.0_dns-settings.lz4
[2022-09-19T19:01:47Z INFO  pubsys::repo::validate_repo] Downloading target: bottlerocket-aws-k8s-1.23-x86_64-v1.9.2.img.lz4
[2022-09-19T19:01:47Z INFO  pubsys::repo::validate_repo] Downloading target: manifest.json
[2022-09-19T19:01:47Z INFO  pubsys::repo::validate_repo] Downloading target: bottlerocket-aws-k8s-1.23-x86_64-1.9.2-deea371c-root.ext4.lz4
[2022-09-19T19:01:47Z INFO  pubsys::repo::validate_repo] Downloading target: bottlerocket-aws-k8s-1.23-x86_64-1.9.2-deea371c-boot.ext4.lz4
[2022-09-19T19:01:47Z INFO  pubsys::repo::validate_repo] Downloading target: bottlerocket-aws-k8s-1.23-x86_64-1.9.2-deea371c.img.lz4
[2022-09-19T19:01:47Z INFO  pubsys::repo::validate_repo] Downloading target: aws-k8s-1.23-x86_64-kmod-kit-v1.9.2.tar.xz
[2022-09-19T19:01:47Z INFO  pubsys::repo::validate_repo] Downloading target: bottlerocket-aws-k8s-1.23-x86_64-1.9.2-deea371c-data.img.lz4
[2022-09-19T19:01:47Z INFO  pubsys::repo::validate_repo] Downloading target: bottlerocket-aws-k8s-1.23-x86_64-v1.9.2-data.img.lz4
[2022-09-19T19:01:47Z INFO  pubsys::repo::validate_repo] Downloading target: bottlerocket-aws-k8s-1.23-x86_64-1.9.2-deea371c-root.verity.lz4
[2022-09-19T19:01:47Z INFO  pubsys::repo::validate_repo] Downloading target: migrate_v1.10.0_dns-settings-metadata.lz4
[cargo-make] INFO - Build Done in 4.25 seconds.

```

Tested `refresh-repo` with KMS key:
```
$ cargo make  -e BUILDSYS_VARIANT=aws-k8s-1.23 -e BUILDSYS_ARCH=x86_64 refresh-repo        
[cargo-make] INFO - cargo make 0.35.15
[cargo-make] INFO - Build File: Makefile.toml
[cargo-make] INFO - Task: refresh-repo
[cargo-make] INFO - Profile: development
[cargo-make] INFO - Running Task: setup
[cargo-make] INFO - Running Task: fetch-sources
[cargo-make] INFO - Running Task: tuftool
[cargo-make] INFO - Running Task: publish-setup-tools
[cargo-make] INFO - Running Task: publish-setup
19:05:28 [INFO] Found infra config at path: /home/ec2-user/bottlerocket/Infra.toml
[cargo-make] INFO - Running Task: publish-tools
[cargo-make] INFO - Running Task: refresh-repo
[2022-09-19T19:05:28Z INFO  pubsys_config] Found infra config at path: /home/ec2-user/bottlerocket/Infra.toml
[2022-09-19T19:05:28Z INFO  pubsys::repo::refresh_repo] Using repo expiration policy from path: /home/ec2-user/bottlerocket/tools/pubsys/policies/repo-expiration/2w-2w-1w.toml
[2022-09-19T19:05:28Z INFO  pubsys::repo::refresh_repo] Loaded TUF repo: file:///home/ec2-user/bottlerocket/build/repos/default/bottlerocket-1.9.2-deea371c-bak/aws-k8s-1.23/x86_64
[2022-09-19T19:05:28Z INFO  pubsys::repo] Setting non-root metadata expiration times:
        snapshot:  2022-10-03 19:05:28.522312124 UTC
        targets:   2022-10-03 19:05:28.522312124 UTC
        timestamp: 2022-09-26 19:05:28.522312124 UTC
[2022-09-19T19:05:28Z INFO  pubsys::repo::refresh_repo] Writing repo metadata to: /home/ec2-user/bottlerocket/build/repos/default/bottlerocket-1.9.2-deea371c/aws-k8s-1.23/x86_64
[cargo-make] INFO - Build Done in 2.40 seconds.
```

bcressey commented on tools/deny.toml:
```diff
@@ -58,14 +58,17 @@ multiple-versions = "deny"
-    # temporarily using a different version of snafu
-    { name = "parse-datetime", version = "0.1.0" },
-
     # rusoto_signature uses an older version of sha2
     { name = "rusoto_signature" },
```
This is still needed until infrasys is updated?

jpculp replied:
Correct.

bcressey commented on tools/pubsys/src/aws/ami/wait.rs:
```diff
@@ -11,8 +11,8 @@ use std::time::Duration;
-    region: &Region,
-    sts_region: &Region,
+    region: &str,
+    sts_region: &str,
```
Why pass `&str` around as the type instead of converting to an `aws_types::region::Region` as early as possible and using that?

bcressey commented on tools/pubsys/src/aws/client.rs:
```diff
@@ -1,159 +1,165 @@
+    S1: AsRef<str>,
+    S2: AsRef<str>,
+{
+    let mut config = aws_config::from_env();
```
Why construct a new config here instead of using the `aws` one that's passed in?

rpkelly replied:
This actually might be really important if the passed in aws config uses a different profile than the default somehow.

jpculp replied:
This aws_config is the SDK's aws_config. Hopefully will be clearer after the rename.

bcressey commented on tools/pubsys/src/aws/client.rs:
```diff
@@ -1,159 +1,165 @@
-}
+pub(crate) async fn build_client_config<S1, S2>(
+    region: S1,
+    sts_region: S2,
```
Why can't these be `Region` types instead?

bcressey commented on tools/pubsys/src/aws/client.rs:
```diff
@@ -1,159 +1,165 @@
+pub(crate) async fn build_client_config<S1, S2>(
+    region: S1,
+    sts_region: S2,
+    aws: &AwsConfig,
```
This should have a more descriptive name:
```suggestion
    pubsys_aws_config: &PubsysAwsConfig,
```

bcressey commented on tools/pubsys/src/aws/client.rs:
```diff
@@ -1,159 +1,165 @@
+use aws_sdk_ec2::{Endpoint, Region};
+use aws_sdk_sts::model::Credentials as StsCredentials;
+use aws_sdk_sts::{Client as StsClient, Credentials as SdkCredentials};
 use pubsys_config::AwsConfig;
```
I found the client builder code very confusing and the ambiguous name of this type was a big part of why.

I would recommend either renaming this type to something like `AwsConfigOverride` or importing it differently:
```suggestion
use pubsys_config::AwsConfig as PubsysAwsConfig;
```

bcressey commented on tools/pubsys/src/aws/client.rs:
```diff
@@ -1,159 +1,165 @@
+    let config = if maybe_regional_role.is_some() {
+        let assume_roles = aws.role.iter().chain(maybe_regional_role.iter()).cloned();
+        let base_config = config.region(Region::new(sts_region)).load().await;
+        build_chained_config(&region, assume_roles.clone(), base_config).await?
```
I may be mis-reading this but it looks like we won't build the chained config unless we have a region-specific role. If there's a top level role in `aws.role` we also need to assume that using the profile credentials, even if there's no regional role.

I.e. this `Infra.toml` would work as expected:
```
[aws]
# If specified, we assume this role before making any API calls.
role = "arn:aws:iam::012345678901:role/assume-global"

[aws.region.us-west-2]
# If specified, we assume this role before making any API calls in this region.
# (This is assumed after the "global" aws.role, if that is also specified.)
role = "arn:aws:iam::012345678901:role/assume-regional"
```

But this one would not:
```
[aws]
# If specified, we assume this role before making any API calls.
role = "arn:aws:iam::012345678901:role/assume-global"
```

We really ought to have unit tests for these if that's possible.

bcressey commented on tools/pubsys/src/aws/client.rs:
```diff
@@ -1,159 +1,165 @@
+            .credentials_provider(credentials.clone())
+            .region(Region::new(region.clone()))
+            .load()
+            .await;
```
Does this end up passing the custom endpoints all the way down the chain? I would guess not because we start with a fresh `aws_config::from_env()` each time.

bcressey commented on tools/pubsys/src/aws/client.rs:
```diff
@@ -1,159 +1,165 @@
+                    .context(error::DateTimeExpirationSnafu)?
+                    .try_into()
+                    .context(error::DateTimeToSystemTimeSnafu)?,
+            ),
```
It seems like if we somehow got `sts_credentials` with no expiration, we should instead pass `None` into this field of `SdkCredentials`, rather than blowing up.

bcressey commented on tools/pubsys/src/aws/mod.rs:
```diff
@@ -10,24 +8,11 @@ pub(crate) mod promote_ssm;
     match input {
-        "x86_64" | "amd64" => Ok("x86_64".to_string()),
-        "arm64" | "aarch64" => Ok("arm64".to_string()),
+        "x86_64" | "amd64" => Ok(ArchitectureValues::X8664),
+        "arm64" | "aarch64" => Ok(ArchitectureValues::Arm64),
```
😬 

Not your fault but the upstream enum and variant names are really unfortunate.

bcressey commented on tools/pubsys/src/aws/promote_ssm/mod.rs:
```diff
@@ -65,10 +65,7 @@ pub(crate) async fn run(args: &Args, promote_args: &PromoteArgs) -> Result<()> {
-    }
-    .into_iter()
-    .map(|name| region_from_string(&name, &aws).context(error::ParseRegionSnafu))
-    .collect::<Result<Vec<Region>>>()?;
```
I don't love dropping the conversion into the more accurate types here at the edges where arguments are parsed. Even if the conversion is trivial and infallible, I'd still prefer to pass around `Region` and `&Region` rather than strings.

bcressey commented on tools/pubsys/src/aws/promote_ssm/mod.rs:
```diff
@@ -78,13 +75,17 @@ pub(crate) async fn run(args: &Args, promote_args: &PromoteArgs) -> Result<()> {
     );
     let base_region = &regions[0];
 
+    let arch = promote_args.arch.as_str();
```
Why convert this back into a string from the more accurate type? That doesn't seem correct.

jpculp replied:
In the SSM modules, `BuildContext` is building a template of **str**-strings, so this was just to simplify code. The AMI modules are the only ones that take advantage of the stronger type. That said, it is probably a good idea to switch to the string at the point that you need it to be one.

bcressey commented on tools/pubsys/src/aws/ami/wait.rs:
```diff
@@ -71,13 +71,18 @@ pub(crate) async fn wait_for_ami(
+                                ImageState::Error
+                            ]
+                            .iter()
+                            .any(|e| e == &found_state),
```
A `match` on the known failure states would be more idiomatic.

bcressey commented on tools/pubsys/src/repo.rs:
```diff
@@ -425,17 +425,33 @@ fn get_signing_key_source(signing_key_config: &SigningKeyConfig) -> Result<Box<d
+    if let Some(region) = kmskey_config.available_keys.get(key_id) {
+        // We are cloning this so that we can send it across a thread boundary
+        let region = region.to_owned();
+        // We need to spin up a new thread to deal with the async nature of the
+        // AWS SDK Rust
+        thread::spawn(move || {
+            let runtime = tokio::runtime::Runtime::new().context(error::RuntimeCreationSnafu)?;
+            Ok(Some(runtime.block_on(async_get_client(&region))))
+        })
+        .join()
+        .map_err(|_| error::Error::ThreadJoin {})?
```
I'd prefer to handle this the same way that `main.rs` handles conditionally creating a Tokio runtime around certain subcommands.

bcressey commented on tools/pubsys/src/aws/publish_ami/mod.rs:
```diff
@@ -348,9 +345,9 @@ pub(crate) async fn modify_snapshots(
-    operation: &str,
-    snapshots: &HashMap<Region, Vec<String>>,
-    clients: &HashMap<Region, Ec2Client>,
+    operation: OperationType,
```
Can this just take a reference instead, so that callers don't have to clone it?
```suggestion
    operation: &OperationType,
```

(Same note for the other functions that now take an `OperationType`.)

bcressey commented on tools/pubsys/src/aws/publish_ami/mod.rs:
```diff
@@ -456,27 +447,30 @@ pub(crate) async fn modify_regional_images(
     for ((region, image_id), modify_image_response) in responses {
         match modify_image_response {
-            Ok(()) => {
+            Ok(_) => {
```
Is there any useful data in the response that's being discarded here? As a hypothetical, if it contained the `image_id`, that would let us simplify the tuple type, or ensure that it matched the expected ID.

jpculp replied:
The response is a [ModifyImageAttributeOutput](https://docs.rs/aws-sdk-ec2/latest/aws_sdk_ec2/output/struct.ModifyImageAttributeOutput.html) type, but I couldn't find a good way to make use of it directly for a success message.

ecpullen commented on tools/pubsys/src/aws/ami/register.rs:
```diff
@@ -74,17 +72,17 @@ async fn _register_image(
-        ..Default::default()
-    };
+    let root_bdm = BlockDeviceMapping::builder()
+        .set_device_name(Some(ROOT_DEVICE_NAME.to_string()))
+        .set_ebs(Some(
+            EbsBlockDevice::builder()
+                .set_delete_on_termination(Some(true))
+                .set_snapshot_id(Some(root_snapshot.clone()))
+                .set_volume_type(Some(VolumeType::from(VOLUME_TYPE)))
+                .set_volume_size(ami_args.root_volume_size)
+                .build(),
```
The sdk allows not including `Some` if you set the values without `set_`. (This happens in a lot of places so might not be in the scope of this pr)
```suggestion
        .device_name(ROOT_DEVICE_NAME.to_string())
        .ebs(EbsBlockDevice::builder()
                .delete_on_termination(true)
                .snapshot_id(root_snapshot.clone())
                .volume_type(VolumeType::from(VOLUME_TYPE))
                .set_volume_size(ami_args.root_volume_size)
                .build(),
```

jpculp replied:
This definitely works, but my understanding was the setters are more idiomatic for the SDK.

bcressey replied:
Where did you see the perspective on idiomatic use? Not disagreeing, and it's not a blocker here in any case, but for me the code that @ecpullen wrote seems preferable since it's fewer characters and we don't have to wrap everything in an `Option`.

jpculp replied:
It was just an impression we got from the SDK team that the non-setter way might eventually be deprecated.

jpculp committed:
pubsys: update to AWS SDK Rust

jpculp commented:
- Brought back stronger-typed regions and `region_from_string()`.
- Renamed `aws: AwsConfig` to `pubsys_aws_config: PubsysAwsConfig` to reduce confusion.
- Kept `arch` as a stronger-type throughout.
- Replaced **ImageState** `ensure` statement with `match`.
- Replaced `thread::spawn` with `tokio::runtime::Runtime::new()`.

bcressey commented on tools/pubsys/src/aws/mod.rs:
```diff
@@ -10,24 +9,16 @@ pub(crate) mod promote_ssm;
-        None => name.parse().context(error::ParseRegionSnafu { name })?,
-    })
+/// Builds a Region from the given region name.
+fn region_from_string(name: &str) -> Result<Region> {
+    Ok(Region::new(name.to_owned()))
```
If this is infallible now we should probably just return a `Region`. However, I like it this way for the purposes of the SDK migration, to minimize the diff. Maybe add a `FIXME` to rework later.

jpculp replied:
This is a pretty quick change. Rather than a `FIXME`, should I just make it the final change once we think this PR hits all the edge cases?

bcressey commented on tools/pubsys/src/aws/client.rs:
```diff
@@ -1,159 +1,169 @@
+) -> Result<SdkConfig> {
+    let mut config = aws_config::from_env();
+
+    // Add custom endpoint if specified in aws.region.REGION.endpoint.
+    let maybe_regional_endpoint = pubsys_aws_config
+        .region
+        .get(region.as_ref())
+        .and_then(|r| r.endpoint.clone());
+    if let Some(endpoint) = &maybe_regional_endpoint {
+        config = config.endpoint_resolver(Endpoint::immutable(
+            endpoint
+                .parse()
+                .context(error::UriParseSnafu { input: endpoint })?,
+        ));
+    }
```
The SDK comment for `endpoint_resolver` says:
```
/// This method will override the endpoint resolver used for **all** AWS services. This mainly                                               
/// exists to set a static endpoint for tools like `LocalStack`. For live traffic, AWS services                                              
/// require the service-specific endpoint resolver they load by default.    
```

Do we actually use that endpoint functionality for anything today? If not, we should just retract the option in `PubsysAwsConfig`.

To handle it correctly, we'd need service-specific endpoint fields in `Infra.toml`, and SDK clients would need to be built to use those endpoints. That would happen in the calling code rather than here.

jpculp replied:
I looked into it, but I wasn't able to find anywhere we are using it today.

bcressey commented on tools/pubsys/src/aws/client.rs:
```diff
@@ -1,159 +1,169 @@
-    }
-}
+use aws_config::{ConfigLoader, SdkConfig};
+use aws_sdk_ec2::{Endpoint, Region};
```
nit: maybe we should use `aws_sdk_sts::Region` here, since that's the only service we end up talking to.

jpculp commented:
- Changed credentials logic will run even if only a single role is specified.
- Credential expirations can now be None.
- Removed endpoints.

zmrow reviewed (approved):
:safety_vest:

jpculp commented:
- Rewrote client.rs to return providers instead of credentials.
- Removed returning unneccessary result-types (and their associated error contexts).

jpculp commented:
Suppressed noisy `[2022-09-22T20:01:26Z INFO  tracing::span] assume_role;` log.

zmrow reviewed (approved):
:microscope:
